text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,249 @@
+* unknown - the size/alignment is not able to be determined at compile time or run time.  
+  This is the category that opaque types fall in (and no other existing types occupy), without any additional domain specific knowledge.
+  Therefore extern types will occupy this category to allow the most flexibility.
+
+The rest of this document will refer to types as "statically sized", "metadata aligned", etc.
+
+"dynamically aligned" (or "unknown aligned") types cannot be placed as the last field of a struct as their offset cannot be determined, without already having a pointer to the field.
+
+In the Rust 2021 edition and earlier `T: Sized` implies that `T` is "statically sized" and "statically aligned" and `T: ?Sized` implies that `T` is "metadata sized" and "metadata aligned".
+
+In the Rust 2024 edition and later `T: Sized` means the same but `T: ?Sized` implies that `T` is "unknown sized" and "unknown aligned".
+A `MetaSized` bound can be introduced to regain the previous meaning.
+
+## `MetaSized` trait
+[metasized-trait]: #metasized-trait
+
+This introduces a new trait `core::marker::MetaSized` that represents a type that is "metadata sized" and "metadata aligned":
+```rust
+#[lang = "meta_sized"]
+trait MetaSized {}
+```
+This trait is automatically implemented for all types except extern types.
+
+All types that implement `MetaSized` can be placed as the last field in a struct because the compiler can emit code to determine the offset of the field purely from the pointer metadata.
+At the time of writing all `MetaSized` types are either statically aligned, or are trait objects and have their alignment in their vtable.
+This RFC does not propose changing codegen for computing the offset as it does not introduce any new `MetaSized` types.
+
+All locations in the standard library that use `?Sized` bounds need to be reviewed when migrating to the 2024 edition and replaced with either `?Sized` or `?Sized + MetaSized` as appropriate. Some examples:
+```rust
+pub fn size_of_val<T: ?Sized + MetaSized>(val: &T) -> usize
+```
+This requires `MetaSized` because the size must be known at run time.
+
+```rust
+pub struct Box<T: ?Sized + MetaSized, A: Allocator = Global>(Unique<T>, A);
+```
+This requires `MetaSized` because the `Box` must be able to determine the Layout of the type at run time to allocate and free the backing memory.
+
+```rust
+impl<A: ?Sized, B: ?Sized> const PartialEq<&B> for &A
+where
+    A: ~const PartialEq<B>,
+```
+This does not require `MetaSized` because the references to `A` and `B` always remain behind a pointer and the size and alignment is not required to call a function on the type.
+
+## Extern types
+[extern-types]: #extern-types
+
+Extern types are defined as above, they are thin DSTs, that is their metadata is `()`. They cannot ever exist except behind a pointer, and so attempts to dereference them fail at compile time similar to trait objects.
```

Comment:
I assume `&*x` is allowed with such types. This means that dereferencing (using the `*` operator) is actually fine, it's only the place-to-value coercion which is forbidden. That matches trait objects (and slices).

Reply:
Yes, exactly. I hadn't considered that distinction, I'll try to clarify that later.